lightning-net-tokio/src/lib.rs

```diff
@@ -1,279 +1,618 @@
+	// are sure we won't call any more read/write PeerManager functions with the same connection.
+	// This is set to true if we're in such a condition (with disconnect checked before with the
+	// top-level mutex held) and false when we can return.
+	block_disconnect_socket: bool,
 	read_paused: bool,
-	need_disconnect: bool,
+	rl_requested_disconnect: bool,
 	id: u64,
 }
 impl Connection {
-	fn schedule_read<CMH: ChannelMessageHandler + 'static>(peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor<CMH>, Arc<CMH>>>, us: Arc<Mutex<Self>>, reader: futures::stream::SplitStream<tokio_codec::Framed<TcpStream, tokio_codec::BytesCodec>>) {
-		let us_ref = us.clone();
-		let us_close_ref = us.clone();
+	fn event_trigger(us: &mut MutexGuard<Self>) {
+		match us.event_notify.try_send(()) {
+			Ok(_) => {},
+			Err(mpsc::error::TrySendError::Full(_)) => {
+				// Ignore full errors as we just need the user to poll after this point, so if they
+				// haven't received the last send yet, it doesn't matter.
+			},
+			_ => panic!()
+		}
+	}
+	async fn schedule_read<CMH: ChannelMessageHandler + 'static>(peer_manager: Arc<peer_handler::PeerManager<SocketDescriptor, Arc<CMH>>>, us: Arc<Mutex<Self>>, mut reader: io::ReadHalf<TcpStream>, mut read_wake_receiver: mpsc::Receiver<()>, mut write_avail_receiver: mpsc::Receiver<()>) {
 		let peer_manager_ref = peer_manager.clone();
-		tokio::spawn(reader.for_each(move |b| {
-			let pending_read = b.to_vec();
-			{
-				let mut lock = us_ref.lock().unwrap();
-				assert!(lock.pending_read.is_empty());
-				if lock.read_paused {
-					lock.pending_read = pending_read;
-					let (sender, blocker) = futures::sync::oneshot::channel();
-					lock.read_blocker = Some(sender);
-					return future::Either::A(blocker.then(|_| { Ok(()) }));
-				}
-			}
-			//TODO: There's a race where we don't meet the requirements of socket_disconnected if its
-			//called right here, after we release the us_ref lock in the scope above, but before we
-			//call read_event!
-			match peer_manager.read_event(&mut SocketDescriptor::new(us_ref.clone(), peer_manager.clone()), pending_read) {
-				Ok(pause_read) => {
-					if pause_read {
-						let mut lock = us_ref.lock().unwrap();
-						lock.read_paused = true;
-					}
-				},
-				Err(e) => {
-					us_ref.lock().unwrap().need_disconnect = false;
-					return future::Either::B(future::result(Err(std::io::Error::new(std::io::ErrorKind::InvalidData, e))));
-				}
+		// 8KB is nice and big but also should never cause any issues with stack overflowing.
+		let mut buf = [0; 8192];
+
+		let mut our_descriptor = SocketDescriptor::new(us.clone());
+		// If Rust-Lightning tells us to disconnect (ie Disconnect::CloseConnection), we don't, then,
+		// call socket_disconnected() (informing Rust-Lightning that we're disconnecting) after
+		// closing the socket, but if the socket closes on its own (ie our counterparty closes the
+		// connection and we set disconnect_type to PeerDisconnected), we do.
```

Comment:
IIUC, rephrase to: 

If Rust-Lightning tells us to disconnect (i.e. through Disconnect::CloseConnection), then we close the socket but
do not call PeerManager's socket_disconnected() (which informs Rust-Lightning that we're disconnecting). But, 
if the socket closes on its own, (i.e. if our counterparty closes the connection, causing us to set disconnect_type
to PeerDisconnected), we do call PeerManager's socket_disconnected().

Reply:
I took Jeff's suggested wording by mixing it into the enum. Is that clearer now?

Reply:
Great wording now!